cli/graph_container.rs

```diff
@@ -155,3 +159,45 @@ impl<'a> ModuleGraphUpdatePermit for MainModuleGraphUpdatePermit<'a> {
     drop(self.permit); // explicit drop for clarity
   }
 }
+
+pub fn resolve_files_from_patterns(
+  pattern_set: PathOrPatternSet,
+) -> Result<Vec<String>, AnyError> {
+  let mut result_files = Vec::new();
+  fn visit_dirs(
+    dir: &Path,
+    pattern_set: &PathOrPatternSet,
+    result_files: &mut Vec<String>,
+  ) -> Result<(), AnyError> {
+    if dir.is_dir() {
+      for entry in fs::read_dir(dir)? {
+        let entry = entry?;
+        let path = entry.path();
+        if path.is_dir() {
+          visit_dirs(&path, pattern_set, result_files)?;
+        } else if pattern_set.matches_path(&path) {
+          result_files.push(path.to_string_lossy().into_owned());
+        }
+      }
+    }
+    Ok(())
+  }
+
+  for pattern in pattern_set.inner() {
+    match pattern {
+      PathOrPattern::Path(path) => {
+        if path.exists() && path.is_file() {
+          result_files.push(path.to_string_lossy().into_owned());
+        }
+      }
+      PathOrPattern::Pattern(pattern) => {
+        let base_path = pattern.base_path();
+        visit_dirs(&base_path, &pattern_set, &mut result_files)?;
+      }
+      PathOrPattern::NegatedPath(_) => {}
+      PathOrPattern::RemoteUrl(url) => result_files.push(url.to_string()),
+    }
+  }
+
+  Ok(result_files)
+}
```

Comment:
We don't need this method. We can use `FileCollector` and pass it the paths and patterns. Right now this code does things like visiting the vendor or node_modules directory. See the lint or fmt subcommands for an example of using `FileCollector`